scipy/signal/_upfirdn_apply.pyx

```diff
@@ -49,40 +51,177 @@ ctypedef fused DTYPE_t:
     double
     double_complex
 
+cdef struct ArrayInfo:
+    np.intp_t * shape
+    np.intp_t * strides
+    np.intp_t ndim
 
-def _output_len(Py_ssize_t len_h,
-                Py_ssize_t in_len,
-                Py_ssize_t up,
-                Py_ssize_t down):
+
+def _output_len(np.intp_t len_h,
+                np.intp_t in_len,
+                np.intp_t up,
+                np.intp_t down):
     """The output length that results from a given input"""
-    cdef Py_ssize_t np
-    cdef Py_ssize_t in_len_copy
+    cdef np.intp_t nt
+    cdef np.intp_t in_len_copy
     in_len_copy = in_len + (len_h + (-len_h % up)) // up - 1
-    np = in_len_copy * up
-    cdef Py_ssize_t need = np // down
-    if np % down > 0:
+    nt = in_len_copy * up
+    cdef np.intp_t need = nt // down
+    if nt % down > 0:
         need += 1
     return need
 
 
-def _apply(DTYPE_t [:] x, DTYPE_t [:] h_trans_flip, DTYPE_t [:] out,
-                 Py_ssize_t up, Py_ssize_t down):
-    _apply_impl(x, h_trans_flip, out, up, down)
+def _apply(np.ndarray data, DTYPE_t [::1] h_trans_flip, np.ndarray out,
+           np.intp_t up, np.intp_t down, np.intp_t axis):
+    cdef ArrayInfo data_info, output_info
+    cdef np.intp_t len_h = h_trans_flip.size
+    cdef DTYPE_t *data_ptr
+    cdef DTYPE_t *filter_ptr
+    cdef DTYPE_t *out_ptr
+    cdef int retval
+
+    # _apply_axis_inner assumes data and out are C-contiguous
+    data = np.ascontiguousarray(data)
+    out = np.ascontiguousarray(out)
+
+    data_info.ndim = data.ndim
+    data_info.strides = <np.intp_t *> data.strides
```

Comment:
@Eric89GXL And because of this we probably do not need the data array to be contiguous. (Though it depends on the logic further downstream.)